naga/src/compact/mod.rs

```diff
@@ -129,62 +203,23 @@ pub fn compact(module: &mut crate::Module) {
         module_tracer.global_expressions_used.insert(constant.init);
     }
 
-    // We treat all named overrides as used by definition.
-    log::trace!("tracing named overrides");
-    for (handle, r#override) in module.overrides.iter() {
-        if r#override.name.is_some() {
-            log::trace!("tracing override {:?}", r#override.name.as_ref().unwrap());
-            module_tracer.overrides_used.insert(handle);
-            module_tracer.types_used.insert(r#override.ty);
-            if let Some(init) = r#override.init {
-                module_tracer.global_expressions_used.insert(init);
+    if preserve_unused {
+        // Treat all named overrides as used.
+        for (handle, r#override) in module.overrides.iter() {
+            if r#override.name.is_some() && !module_tracer.overrides_used.contains(handle) {
+                module_tracer.overrides_used.insert(handle);
+                module_tracer.types_used.insert(r#override.ty);
+                if let Some(init) = r#override.init {
+                    module_tracer.global_expressions_used.insert(init);
+                }
             }
         }
-    }
-
-    // We assume that all functions are used.
-    //
-    // Observe which types, constant expressions, constants, and
-    // expressions each function uses, and produce maps for each
-    // function from pre-compaction to post-compaction expression
-    // handles.
-    log::trace!("tracing functions");
-    let function_maps: Vec<FunctionMap> = module
-        .functions
-        .iter()
-        .map(|(_, f)| {
-            log::trace!("tracing function {:?}", f.name);
-            let mut function_tracer = module_tracer.as_function(f);
-            function_tracer.trace();
-            FunctionMap::from(function_tracer)
-        })
-        .collect();
 
-    // Similarly, observe what each entry point actually uses.
-    log::trace!("tracing entry points");
-    let entry_point_maps: Vec<FunctionMap> = module
-        .entry_points
-        .iter()
-        .map(|e| {
-            log::trace!("tracing entry point {:?}", e.function.name);
-
-            if let Some(sizes) = e.workgroup_size_overrides {
-                for size in sizes.iter().filter_map(|x| *x) {
-                    module_tracer.global_expressions_used.insert(size);
-                }
+        // With `preserve_unused`, treat all named types as used.
```

Comment:
suggestion: The context of `preserve_unused` is so close, it seems redundant to even mention it:

```suggestion
        // Treat all named types as used.
```

Reply:
Pushed as a `fixup!` with 7e99152d.